src/passes/Asyncify.cpp

```diff
@@ -683,28 +712,31 @@ class ModuleAnalyzer {
           canChangeState = true;
         }
       }
-      void visitCallIndirect(CallIndirect* curr) {
-        if (canIndirectChangeState) {
-          canChangeState = true;
-        }
-        // TODO optimize the other case, at least by type
-      }
+      void visitCallIndirect(CallIndirect* curr) { hasIndirectCall = true; }
       Module* module;
       ModuleAnalyzer* analyzer;
       Map* map;
-      bool canIndirectChangeState;
+      bool hasIndirectCall = false;
       bool canChangeState = false;
       bool isBottomMostRuntime = false;
     };
     Walker walker;
     walker.module = &module;
     walker.analyzer = this;
     walker.map = &map;
-    walker.canIndirectChangeState = canIndirectChangeState;
     walker.walk(curr);
     if (walker.isBottomMostRuntime) {
       walker.canChangeState = false;
     }
+    // An indirect call is normally ignored if we are ignoring indirect calls.
+    // However, see the docs at the top: if the function we are inside was
+    // specifically added by the user (in the only-list or the add-list) then we
+    // instrument indirect calls from it (this allows specifically allowing some
+    // indirect calls but not others).
+    if (walker.hasIndirectCall &&
+        (canIndirectChangeState || map[func].addedFromList)) {
+      walker.canChangeState = true;
+    }
```

Comment:
It looks like this logic could be combined with the `return` below to be just one return statement. I think this might be simpler than the the current code.

Reply:
How are you suggesting they be combined?

I guess 738 could be `return true`. Or the line could be `return walker.canChangeState || (walker.hasIndirectCall && (canIndirectChangeState || map[func].addedFromList))`. But both seem less good to me?

Reply:
I was thinking the latter, but I think the former would also be an improvement. I just prefer to have to track less state and control flow. But this is very much a nit, so I leave it to you :)

Reply:
Just to explain why I prefer it as it is, combined with the lines above, this overall code has the form
```
compute result
// explain exception 1
if ..exception1.. set result to false
// explain exception 2
if ..complex exception2.. set result to true
return result
```
Actually re-reading it now, those two exceptions are in the wrong order anyhow, so this is moot... reordering, and then the condition before the return is very small so I agree it can be merged with the return, doing it that way.